src/main/kotlin/pl/edu/uj/ii/ksi/mordor/controllers/FilesystemController.kt

```diff
@@ -149,6 +155,25 @@ class FilesystemController(
         return ModelAndView(RedirectView(urlEncodePath("/download/${entity.relativePath}")))
     }
 
+    @Secured(Permission.READ_STR)
+    @GetMapping("/thumbnail/**")
+    fun fileThumbnail(request: HttpServletRequest, response: HttpServletResponse) {
+        val path = request.servletPath.removePrefix("/thumbnail")
+        val entity = (repoService.getEntity(path)
+                ?: throw NotFoundException(path)) as? RepositoryFile
+                ?: throw BadRequestException("not a file")
+
+        val thumbnail = entryRepository.findById(path).get().metadata?.thumbnail?.thumbnail
+        response.contentType = entity.browserSafeMimeType
```

Comment:
Return type of the thumbnail itself, not the entity.
```suggestion
        response.contentType = "image/png"
```